src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -633,7 +633,14 @@ impl<'a, 'tcx> Visitor<'tcx> for Qualifier<'a, 'tcx, 'tcx> {
 
                 // We might have a candidate for promotion.
                 let candidate = Candidate::Ref(location);
-                if !self.qualif.intersects(Qualif::NEVER_PROMOTE) {
+                let mut unpromotable = self.qualif & Qualif::NEVER_PROMOTE;
```

Comment:
nit: I would prefer that this codepath be shared with the shuffle index code into a `fn can_promote`